src/Fhir.Anonymizer.Core/Visitors/AnonymizationVisitor.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Text;
+using Fhir.Anonymizer.Core.AnonymizationConfigurations;
+using Fhir.Anonymizer.Core.Extensions;
+using Fhir.Anonymizer.Core.Models;
+using Fhir.Anonymizer.Core.Processors;
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Specification;
+using Hl7.FhirPath;
+
+namespace Fhir.Anonymizer.Core.Visitors
+{
+    public class AnonymizationVisitor : AbstractElementNodeVisitor
+    {
+        private AnonymizationFhirPathRule[] _rules;
+        private Dictionary<string, IAnonymizerProcessor> _processors;
+        private HashSet<ElementNode> _visitedNodes = new HashSet<ElementNode>();
+        private Stack<Tuple<ElementNode, ProcessResult>> _contextStack = new Stack<Tuple<ElementNode, ProcessResult>>();
+
+        public bool AddSecurityTag { get; set; } = true;
+
+        public AnonymizationVisitor(AnonymizationFhirPathRule[] rules, Dictionary<string, IAnonymizerProcessor> processors)
+        {
+            _rules = rules;
+            _processors = processors;
+        }
+
+        public override bool Visit(ElementNode node)
+        {
+            if (node.IsFhirResource())
+            {
+                ProcessResult result = ProcessResourceNode(node);
+                _contextStack.Push(new Tuple<ElementNode, ProcessResult>(node, result));
+            }
+
+            return true;
+        }
+
+        public override void EndVisit(ElementNode node)
+        {
+            if (node.IsFhirResource())
+            {
+                Tuple<ElementNode, ProcessResult> context = _contextStack.Pop();
+                ProcessResult result = context.Item2;
+
+                if (context.Item1 != node)
+                {
+                    throw new ConstraintException("Internal error: access wrong context.");
```

Comment:
When will we throw this exception?

Reply:
yes, it should never throw. Add check in case any bug in the traverse and fail fast.